### ISSUE TITLE
How do you define a custom typed header?

### ISSUE THREAD

**Issue by kevinastone** [hyperium/headers#29]
I'm trying to define a typed header for `X-Forward-For`, but I can't get around the `const NAME: &'static HeaderName` requirement which only works with the standard headers (to my knowledge).

Could/should this switch the `Header` trait to a name() method instead of an inner constant?

**Comment by kevinastone** [comment]
I'd like to have something closer to:
```
lazy_static! {
    static ref X_FORWARD_FOR: HeaderName =
        HeaderName::from_static("X-Forwarded-For");
}

#[derive(Clone, Debug, PartialEq)]
pub struct XForwardFor(pub String);

impl Header for XForwardFor {
    fn name() -> HeaderName {
        X_FORWARD_FOR
    }

    [...]
}
```

**Comment by seanmonstar** [comment]
Very soon (supposedly), with expanded functionality of `const fn`, we should be able to provide a `const fn` to create a `HeaderName`, which should allow for creating *new* names that can be constants.

Since that is coming, my preference is to not have to change to a `fn name` on the trait. How does that sound to you?

**Comment by kevinastone** [comment]
I looked into `const fn`.  I don't think `HeaderName::from_static` would comply with a compile time function.  Do we intend to rewrite these `from_static` helpers to only use `const fn` themselves?

**Comment by davidbarsky** [comment]
@kevinastone I've landed a PR in master with #34, based on your suggested changes. I think the next step would be to make the derive public with a _mandatory_ attribute that points to a header name. I'm thinking of something along these lines:

```rust
use lazy_static::lazy_static;
use std::net::IpAddr;

lazy_static! {
    pub static ref X_FORWARD_FOR: HeaderName = HeaderName::from_static("x-forwarded-for");
}

#[derive(Header)]
#[header(name = &X_FORWARD_FOR)]
struct ForwardFor(Vec<IpAddr>)
```

(I'm starting to see why @seanmonstar wanted better `const fn`s. Ah well.)

Thoughts before I submit a PR?

**Comment by seanmonstar** [comment]
Thanks to #34, this should now be possible.
